### ISSUE TITLE
CCS against '*:index_name' only returns results if index_name is in all clusters

### ISSUE THREAD

**Issue by rtkjliviero** [elastic/elasticsearch#45010]
# Bug Report
## Elasticsearch version
7.2.0, using the [Docker image provided by Elastic](https://www.elastic.co/guide/en/elasticsearch/reference/7.2/docker.html):
```
$ docker images | grep elasticsearch
REPOSITORY                                                             TAG                 IMAGE ID            CREATED             SIZE
docker.elastic.co/elasticsearch/elasticsearch                          7.2.0               0efa6a3de177        5 weeks ago         861MB
```
## Plugins installed
Basing this on the output from `docker-compose up`:
* aggs-matrix-stats
* analysis-common
* data-frame
* ingest-common
* ingest-geoip
* ingest-user-agent
* lang-expression
* lang-mustache
* lang-painless
* mapper-extras
* parent-join
* percolator
* rank-eval
* reindex
* repository-url
* transport-netty4
* x-pack-ccr
* x-pack-core
* x-pack-deprecation
* x-pack-graph
* x-pack-ilm
* x-pack-logstash
* x-pack-ml
* x-pack-monitoring
* x-pack-rollup
* x-pack-security
* x-pack-sql
* x-pack-watcher

...But I'll defer to "whatever is in that docker image"

## JVM version:
`JVM[Oracle Corporation/OpenJDK 64-Bit Server VM/12.0.1/12.0.1+12]` on all nodes.  Again this is based on output, but I haven't touched the docker image at all.
From inside one of the containers:
```
[root@fbdb3a975ac9 elasticsearch]#  /usr/share/elasticsearch/jdk/bin/java -version
openjdk version "12.0.1" 2019-04-16
OpenJDK Runtime Environment (build 12.0.1+12)
OpenJDK 64-Bit Server VM (build 12.0.1+12, mixed mode, sharing)
```

## OS version
Host is `Darwin C02XM437JGH6 18.7.0 Darwin Kernel Version 18.7.0`
Containers are `Linux fbdb3a975ac9 4.9.125-linuxkit #1 SMP Fri Sep 7 08:20:28 UTC 2018 x86_64 x86_64 x86_64 GNU/Linux`, or in more detail:
```
[root@fbdb3a975ac9 elasticsearch]# cat /etc/os-release
NAME="CentOS Linux"
VERSION="7 (Core)"
ID="centos"
ID_LIKE="rhel fedora"
VERSION_ID="7"
PRETTY_NAME="CentOS Linux 7 (Core)"
ANSI_COLOR="0;31"
CPE_NAME="cpe:/o:centos:centos:7"
HOME_URL="https://www.centos.org/"
BUG_REPORT_URL="https://bugs.centos.org/"

CENTOS_MANTISBT_PROJECT="CentOS-7"
CENTOS_MANTISBT_PROJECT_VERSION="7"
REDHAT_SUPPORT_PRODUCT="centos"
REDHAT_SUPPORT_PRODUCT_VERSION="7"
```

## Description of the problem including expected versus actual behavior:
Cross-cluster search results against 'wildcard' cluster are inconsistent 
1. Performing an index search against `*:index_from_one_cluster` returns no results (`index_not_found_exception`), but specifying the cluster (`cluster_two:index_name`) does
2. Performing an index search against `*:index_from_both_clusters`, which lives on a different cluster, works as expected - results are returned from both clusters.
This is problematic as we are trying to leverage CCS to locate given indices within several clusters.

## Steps to reproduce:
 1. `docker-compose up` with the following docker-compose.yaml file:
```
version: '2.3'
services:
  es7-01:
    image: docker.elastic.co/elasticsearch/elasticsearch:7.2.0
    container_name: es7-01
    environment:
      - node.name=es7-01
      - cluster.initial_master_nodes=es7-01
      - cluster.name=docker-cluster-es7-1
      - bootstrap.memory_lock=true
      - "ES_JAVA_OPTS=-Xms128m -Xmx128m"
    ulimits:
      memlock:
        soft: -1
        hard: -1
    volumes:
      - es7data01:/usr/share/elasticsearch7/data
    ports:
      - 9271:9200
    networks:
      - es7net
    healthcheck:
      test: ["CMD", "curl", "-f", "http://localhost:9200"]
      interval: 40s
      timeout: 5s
      retries: 3
      start_period: 40s
  es7-02:
    image: docker.elastic.co/elasticsearch/elasticsearch:7.2.0
    container_name: es7-02
    environment:
      - node.name=es7-02
      - cluster.initial_master_nodes=es7-02
      - cluster.name=docker-cluster-es7-2
      - bootstrap.memory_lock=true
      - "ES_JAVA_OPTS=-Xms128m -Xmx128m"
    ulimits:
      memlock:
        soft: -1
        hard: -1
    volumes:
      - es7data02:/usr/share/elasticsearch7/data
    ports:
      - 9272:9200
    networks:
      - es7net
    healthcheck:
      test: ["CMD", "curl", "-f", "http://localhost:9200"]
      interval: 40s
      timeout: 5s
      retries: 3
      start_period: 40s
  es7-coord:
    depends_on:
      - es7-01
      - es7-02
    image: docker.elastic.co/elasticsearch/elasticsearch:7.2.0
    container_name: es7-coord
    environment:
      - node.name=es7-coord
      - cluster.initial_master_nodes=es7-coord
      - cluster.name=docker-cluster-es7-coord
      - bootstrap.memory_lock=true
      - "ES_JAVA_OPTS=-Xms128m -Xmx128m"
      - search.remote.cluster_one.seeds=es7-01:9300
      - search.remote.cluster_two.seeds=es7-02:9300
    ulimits:
      memlock:
        soft: -1
        hard: -1
    volumes:
      - es7datacoord:/usr/share/elasticsearch7/data
    ports:
      - 9273:9200
    networks:
      - es7net
    healthcheck:
      test: ["CMD", "curl", "-f", "http://localhost:9200"]
      interval: 40s
      timeout: 5s
      retries: 3
      start_period: 40s
volumes:
  es7data01:
    driver: local
  es7data02:
    driver: local
  es7datacoord:
    driver: local

networks:
  es7net:
```
 2. Load trivial sample data with the following Python snippet:
```
class TestClusterInfo(object):
  """ Note that in this set of tests, we are only using single-node clusters"""
  def __init__(self, cluster_name, node_host, port, expected_data):
    self.name = cluster_name
    self.node_host = node_host
    self.port = port  # Note: host port!
    self.expected_data = expected_data
    self.host = 'localhost'

COMMON_CLUSTER_DATA = {
  'colours': [
    'blue',
    'red',
    'yellow'
  ]
}

CLUSTER_ONE_DATA = dict({
  'shapes': [
    'circle',
    'square',
    'triangle'
  ]
}, **COMMON_CLUSTER_DATA)

CLUSTER_TWO_DATA = dict({
  'animals': [
    'bears',
    'lions',
    'tigers'
  ]
}, **COMMON_CLUSTER_DATA)

SINGLE_CLUSTERS = [
  TestClusterInfo('docker-cluster-es7-1', 'es7-01', 9271, CLUSTER_ONE_DATA),
  TestClusterInfo('docker-cluster-es7-2', 'es7-02', 9272, CLUSTER_TWO_DATA)
]

def _prepopulate_test_data():
  import elasticsearch
  for cluster in SINGLE_CLUSTERS:
    host = {'host': cluster.host, 'port': cluster.port}
    conn = elasticsearch.Elasticsearch(hosts=[host])
    for index, docs in cluster.expected_data.items():
      for num, doc in enumerate(docs):
        conn.create(
          index=index,
          id=num,
          doc_type='_doc',
          body={doc: {}},
        )
```
 3. Our indices loaded as expected:
```
$ curl 'http://localhost:9271/_cat/indices'  # first cluster
yellow open shapes  IiQAHAc2QUSXt2F56hqDHw 1 1 3 0 2.7kb 2.7kb
yellow open colours KN8v6lNnSCmgum7ttsp2Kw 1 1 3 0 2.7kb 2.7kb
$ curl 'http://localhost:9272/_cat/indices'  # second cluster
yellow open animals QISHifjRQvqYoWXps0ebPw 1 1 3 0 2.7kb 2.7kb
yellow open colours 2DarRiu5SG-KwLoijrGGyQ 1 1 3 0 2.7kb 2.7kb
$ curl 'http://localhost:9273/_cat/indices'  # CCS "coordindator" node
$
```
4. We can find the common index in both clusters with a wildcard search:
```
$ curl 'http://localhost:9273/*:colours/_search' -H 'Content-Type: application/json' -d '{"query": {"match_all": {}}}'
{"took":9,"timed_out":false,"num_reduce_phases":3,"_shards":{"total":2,"successful":2,"skipped":0,"failed":0},"_clusters":{"total":2,"successful":2,"skipped":0},"hits":{"total":{"value":6,"relation":"eq"},"max_score":1.0,"hits":[{"_index":"cluster_two:colours","_type":"_doc","_id":"0","_score":1.0,"_source":{"blue":{}}},{"_index":"cluster_two:colours","_type":"_doc","_id":"1","_score":1.0,"_source":{"red":{}}},{"_index":"cluster_two:colours","_type":"_doc","_id":"2","_score":1.0,"_source":{"yellow":{}}},{"_index":"cluster_one:colours","_type":"_doc","_id":"0","_score":1.0,"_source":{"blue":{}}},{"_index":"cluster_one:colours","_type":"_doc","_id":"1","_score":1.0,"_source":{"red":{}}},{"_index":"cluster_one:colours","_type":"_doc","_id":"2","_score":1.0,"_source":{"yellow":{}}}]}}
```
5. But _not_ either index which appears in only a single cluster:
```
$ curl 'http://localhost:9273/*:shapes/_search' -H 'Content-Type: application/json' -d '{"query": {"match_all": {}}}'
{"error":{"root_cause":[{"type":"index_not_found_exception","reason":"no such index [shapes]","index_uuid":"_na_","resource.type":"index_or_alias","resource.id":"shapes","index":"shapes"}],"type":"index_not_found_exception","reason":"no such index [shapes]","index_uuid":"_na_","resource.type":"index_or_alias","resource.id":"shapes","index":"shapes"},"status":404}
$
$
$ curl 'http://localhost:9273/*:animals/_search' -H 'Content-Type: application/json' -d '{"query": {"match_all": {}}}'
{"error":{"root_cause":[{"type":"index_not_found_exception","reason":"no such index [animals]","index_uuid":"_na_","resource.type":"index_or_alias","resource.id":"animals","index":"animals"}],"type":"index_not_found_exception","reason":"no such index [animals]","index_uuid":"_na_","resource.type":"index_or_alias","resource.id":"animals","index":"animals"},"status":404}
```
6. We can get results by being specific about the index:
```
$ curl 'http://localhost:9273/cluster_one:shapes/_search' -H 'Content-Type: application/json' -d '{"query": {"match_all": {}}}'
{"took":5,"timed_out":false,"_shards":{"total":1,"successful":1,"skipped":0,"failed":0},"_clusters":{"total":1,"successful":1,"skipped":0},"hits":{"total":{"value":3,"relation":"eq"},"max_score":1.0,"hits":[{"_index":"cluster_one:shapes","_type":"_doc","_id":"0","_score":1.0,"_source":{"circle":{}}},{"_index":"cluster_one:shapes","_type":"_doc","_id":"1","_score":1.0,"_source":{"square":{}}},{"_index":"cluster_one:shapes","_type":"_doc","_id":"2","_score":1.0,"_source":{"triangle":{}}}]}}
$
$
$ curl 'http://localhost:9273/cluster_two:animals/_search' -H 'Content-Type: application/json' -d '{"query": {"match_all": {}}}'
{"took":6,"timed_out":false,"_shards":{"total":1,"successful":1,"skipped":0,"failed":0},"_clusters":{"total":1,"successful":1,"skipped":0},"hits":{"total":{"value":3,"relation":"eq"},"max_score":1.0,"hits":[{"_index":"cluster_two:animals","_type":"_doc","_id":"0","_score":1.0,"_source":{"bears":{}}},{"_index":"cluster_two:animals","_type":"_doc","_id":"1","_score":1.0,"_source":{"lions":{}}},{"_index":"cluster_two:animals","_type":"_doc","_id":"2","_score":1.0,"_source":{"tigers":{}}}]}}
```
...But that's not helpful if trying to locate a specific index.

**Comment by ywelsch** [comment]
I've relabeled this as search/search as I think this particular issue is better handled by that team (in particular @javanna).

**Comment by javanna** [comment]
hi @rtkjliviero , `*:index_from_one_cluster` will be resolved by the coordinating node to `cluster1:index_from_one_cluster,cluster2:index_from_one_cluster,cluster3:index_from_one_cluster` (assuming you have those clusters registered, just an example). By default, referring explicitly to indices that don't exists yields an error though, try with a local cluster (no CCS): `GET /does_not_exist/_search` and you should see the same index not found exception.

Can you try and provide `ignore_unavailable=true&allow_no_indices=true` to your search request please?

**Comment by rtkjliviero** [comment]
Thanks @javanna, providing those arguments to my search request works!  Presumably I can pass the same args in the Python API?  That's what I eventually want to be using - using `curl` was just for testing/error reproduction.

Thanks also for the explanation - with the wildcard expansion, I can see how the behaviour I'm seeing would occur.

Still, this doesn't seem very intuitive as a user.  I would expect `*` to imply "find this index _wherever_ it might be," especially because it works that way when the index _does_ exist in all clusters.  Would there be value in tweaking the CCS behaviour in that way - basically make `ignore_unavailable` and `allow_no_indices` default to true?

**Comment by javanna** [comment]
glad to hear @rtkjliviero ! you should be able to provide these options using any of our clients. 

> Still, this doesn't seem very intuitive as a user. I would expect * to imply "find this index wherever it might be," especially because it works that way when the index does exist in all clusters. Would there be value in tweaking the CCS behaviour in that way - basically make ignore_unavailable and allow_no_indices default to true?

I see what you mean, yet I am not sure we'd want to change default settings for CCS searches,  as  they would then differ from ordinary search defaults. And I think that the default behaviour for search is ok and well accepted by our users: throw error when an index that is explicitly mentioned is not found.

**Comment by javanna** [comment]
I will mark team-discuss though so we can have a short discussion about it and gather opinions from the rest of the team.

**Comment by javanna** [comment]
We have discussed this and we have concluded that it would be nice to change the behaviour when CCS is used, so that we send each remote request being lenient and delay throwing an exception in the coordinating node only in case the indices was not found anywhere. There may edge cases so we may find this harder to implement than it seems at first, but it's worth having a look at it. I will mark this adopt-me.
